tracing/src/main/java/com/palantir/remoting3/tracing/Tracers.java

```diff
@@ -177,19 +171,61 @@ public V call() throws Exception {
      */
     private static class TracingAwareRunnable implements Runnable {
         private final Runnable delegate;
-        private final Trace trace;
+        private DeferredTracer deferredTracer;
 
         TracingAwareRunnable(Runnable delegate) {
             this.delegate = delegate;
-            this.trace = Tracer.copyTrace();
+            this.deferredTracer = new DeferredTracer();
         }
 
         @Override
         public void run() {
+            deferredTracer.withTrace(() -> {
+                delegate.run();
+                return null;
+            });
+        }
+    }
+
+    public interface ThrowingCallable<T, E extends Throwable> {
+        T call() throws E;
+    }
+
+    /**
+     * Utility class for capturing the current trace at time of construction, and then
+     * running callables at some later time with that captured trace.
+     * <pre>
+     * <code>
+     * DeferredTracer deferredTracer = new DeferredTracer();
+     *
+     * //...
+     *
+     * // some time later
+     * deferredTracer.withTrace(() -> {
+     *     doThings();
+     *     System.out.println(Tracer.getTraceId()); // prints trace id at time of construction of deferred tracer
+     *     return null;
+     * });
+     *
+     * </code>
+     * </pre>
+     */
+    public static final class DeferredTracer {
+        private final Trace trace;
+
+        public DeferredTracer() {
+            this.trace = Tracer.copyTrace();
```

Comment:
can we inject the trace here rather than grabbing it from static state?

Reply:
the whole point is that the static state is private

Reply:
...this avoids making Tracer.copyTrace() world public

Reply:
oh, i see. cool.